drivers/can/mcp2515.c

```diff
@@ -0,0 +1,748 @@
+	config_buf[3] = caninte;
+
+	mcp2515_write_reg(dev, MCP2515_ADDR_CNF3, config_buf, 4);
+
+	/* Receive everything, filtering done in driver, RXB0 roll over into RXB1 */
+	const u8_t rx0_ctrl = BIT(6) | BIT(5) | BIT(2);
+	const u8_t rx1_ctrl = BIT(6) | BIT(5);
+
+	mcp2515_bit_modify(dev, MCP2515_ADDR_RXB0CTRL, rx0_ctrl, rx0_ctrl);
+	mcp2515_bit_modify(dev, MCP2515_ADDR_RXB1CTRL, rx1_ctrl, rx1_ctrl);
+
+	return mcp2515_request_opmode(dev, mcp2515_get_mcp2515_opmode(mode));
+}
+
+int mcp2515_send(struct device *dev, struct can_msg *msg, s32_t timeout,
+		 can_tx_callback_t callback)
+{
+	int tx_idx = 0;
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	if (k_sem_take(&dev_data->tx_sem, timeout) != 0) {
+		return CAN_TIMEOUT;
+	}
+
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
+	/* find a free tx slot */
+	for (; tx_idx < MCP2515_TX_CNT; tx_idx++) {
+		if ((BIT(tx_idx) & dev_data->tx_busy_map) == 0) {
+			dev_data->tx_busy_map |= BIT(tx_idx);
+			break;
+		}
+	}
+	k_mutex_unlock(&dev_data->tx_mutex);
+
+	if (tx_idx == MCP2515_TX_CNT) {
+		SYS_LOG_WRN("no free tx slot available");
+		return CAN_TX_ERR;
+	}
+
+	dev_data->tx_cb[tx_idx].cb = callback;
+
+	u8_t addr_tx = MCP2515_ADDR_TXB0CTRL + (tx_idx * 0x10);
+	u8_t tx_frame[MCP2515_FRAME_LEN];
+
+	mcp2515_convert_can_msg_to_mcp2515_frame(msg, tx_frame);
+	mcp2515_write_reg(dev, addr_tx + 1, tx_frame, MCP2515_FRAME_LEN);
+	/* request tx slot transmission */
+	mcp2515_bit_modify(dev, addr_tx, BIT(3), BIT(3));
+
+	if (callback == NULL) {
+		k_sem_take(&dev_data->tx_cb[tx_idx].sem, K_FOREVER);
+		return 0;
+	}
+
+	return 0;
+}
+
+int mcp2515_attach_msgq(struct device *dev, struct k_msgq *msgq,
+			const struct can_filter *filter)
+{
+	return mcp2515_attach(dev, filter, (void *) msgq, 1);
+}
+
+int mcp2515_attach_isr(struct device *dev, can_rx_callback_t isr,
+		       const struct can_filter *filter)
+{
+	return mcp2515_attach(dev, filter, (void *) isr, 0);
+}
+
+void mcp2515_detach(struct device *dev, int filter_nr)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	k_mutex_lock(&dev_data->filter_mutex, K_FOREVER);
+	dev_data->filter_usage &= ~BIT_LL(filter_nr);
+	k_mutex_unlock(&dev_data->filter_mutex);
+}
+
+static u8_t mcp2515_filter_match(struct can_msg *msg, struct can_filter *filter)
+{
+	if (msg->id_type != filter->id_type) {
+		return 0;
+	}
+	if ((msg->rtr ^ filter->rtr) & filter->rtr_mask) {
+		return 0;
+	}
+
+	if (msg->id_type == CAN_STANDARD_IDENTIFIER) {
+		if ((msg->std_id ^ filter->std_id) & filter->std_id_mask) {
+			return 0;
+		}
+	} else {
+		if ((msg->ext_id ^ filter->ext_id) & filter->ext_id_mask) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static void mcp2515_rx_filter(struct device *dev, struct can_msg *msg)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+	u8_t filter_idx = 0;
+
+	k_mutex_lock(&dev_data->filter_mutex, K_FOREVER);
+
+	for (; filter_idx < CONFIG_CAN_MAX_FILTER; filter_idx++) {
+		if (BIT_LL(filter_idx) & dev_data->filter_usage) {
```

Comment:
too many if imbrications.

I guess you could reverse that test and use continue, to reduce indentation of the lines below.
If necessary, it's also a good idea to create an inline function to de-clutter the code.

Reply:
Done